### PR TITLE
Refactor align

### PR DESCRIPTION
**Pull request by kain88-de**
It stills needs some checking and tests. I would like go get this into 0.16.0 though. The RMSD class has been refactored this cycle and I'm currently not sure if the bug has been present as well in 0.15.0

Changes made in this Pull Request:
 - use arbitrary weights in align.py and rms.py
 - rewrite RMSF to use AnalysisBase
 - fix centering bug in RMSD (should be done with chosen weights)

PR Checklist
------------
 - [x] Tests?
 - [x] tests for new deprecated option
 - [x] Docs?
 - [x] CHANGELOG updated?
 - ~~[ ] Issue raised/referenced?~~
 - [x] add deprecation warnings for `mass_weighted`
 - [x] updated functions depending on new kwarg
 - [ ] fix all mass_weighted in encore. Or at least raise an issue and fix in other PR

### CODE REVIEW

**Comment by kain88-de** [comment]
@mtiberti @wouterboomsma I see that encore also uses the `mass_weighted` argument. Would it be hard to change it to  `weights='mass'` to allow for arbitrary weights like I introduce here in the PR for rms and alignment function?

**Comment by kain88-de** [comment]
Ok So i'm think I'm done with this. All the internal references to the old keyword have been replaced and appropriate deprecation messages are raised right now when the old one is used. Only the encore module still use the old keyword. I haven't looked yet at how hard it would be to change them as well.

**Comment by kain88-de** [comment]
I haven't looked into actually using arbitrary weights in the tests. I will still do that.

**Comment by kain88-de** [comment]
@MDAnalysis/coredevs can anyone see why the travis build fails. For some reason nosetest doesn't exit with 0.

File: testsuite/MDAnalysisTests/analysis/test_rms.py [hunk 1]
```diff
@@ -183,7 +188,19 @@ def test_rmsd_single_frame(self):
...
+        assert_array_almost_equal(RMSD.rmsd, saved, 4,
+                                  err_msg="error: rmsd profile should match " +
+                                          "test values")
+
+    def test_mass_weighted_and_save_deprecated(self):
+        with warnings.catch_warnings(record=True) as warn:
+            warnings.simplefilter('always')
+            RMSD = MDAnalysis.analysis.rms.RMSD(self.universe, select='name CA',
+                                                step=49, mass_weighted=True).run()
+        assert(len(warn), 1)
```
Comment (tylerjereddy) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: Travis CI build seems to whine about this with `testsuite/MDAnalysisTests/analysis/test_rms.py:203: [W0199(assert-on-tuple), TestRMSD.test_mass_weighted_and_save_deprecated] Assert called on a 2-uple. Did you mean 'assert x,y'?`

Comment (kain88-de) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: thanks. I meant `assert_equal`

**Comment by kain88-de** [comment]
OK I'm done now this PR. Someone could please review this. Their is still the the `mass_weighted` kwarg in encore. But to remove that I will do some refactoring independent of this PR to simplify the code.

**Comment by kain88-de** [comment]
How can I find out in the coverage API which lines I have missed now with the changes here.

**Comment by richardjgowers** [comment]
https://coveralls.io/builds/9478683 this shows the current coverage for this commit.  coveralls doesn't let you see the delta in lines caused by this PR though, (I think codecov does though)

File: package/MDAnalysis/analysis/align.py [hunk 2]
```diff
@@ -339,57 +339,58 @@ def alignto(mobile, reference, select="all", mass_weighted=False,
...
+           Apply to `mobile.select_atoms(selection-string)`
+       :class:`~MDAnalysis.core.groups.AtomGroup`
+           Apply to the arbitrary group of atoms
 
     Returns
     -------
     old_rmsd
         RMSD before spatial alignment
     new_rmsd
-        RMSD after spatial alignment.
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/align.py]: Why did you remove the description?

File: package/MDAnalysis/analysis/align.py [hunk 3]
```diff
@@ -404,6 +405,9 @@ def alignto(mobile, reference, select="all", mass_weighted=False,
        Uses :func:`get_matching_atoms` to work with incomplete selections
        and new *strict* keyword. The new default is to be lenient whereas
        the old behavior was the equivalent of *strict* = ``True``.
+
+    .. versionchanged:: 0.16.0
+       new general 'weights' kwarg replace mass_weights
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/align.py]: add deprecation for `mass_weighted` kw

File: package/MDAnalysis/analysis/align.py [hunk 4]
```diff
@@ -339,57 +339,58 @@ def alignto(mobile, reference, select="all", mass_weighted=False,
...
+          produces identical selections in *mobile* and *reference*; or
+       2. dictionary ``{'mobile':sel1, 'reference':sel2}``.
+          (the :func:`fasta2select` function returns such a
+          dictionary based on a ClustalW_ or STAMP_ sequence alignment); or
+       3.  tuple ``(sel1, sel2)``
+
+       When using 2. or 3. with *sel1* and *sel2* then these selections can
+       also each be a list of selection strings (to generate a AtomGroup with
+       defined atom order as described under :ref:`ordered-selections-label`).
+    mass_weighted : boolean, optional, deprecated
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/align.py]: below you wrote this as `(deprecated) – consistency?

File: package/MDAnalysis/analysis/encore/covariance.py [hunk 5]
```diff
@@ -173,41 +173,34 @@ def shrinkage_covariance_estimator( coordinates,
 def covariance_matrix(ensemble,
                       selection="name CA",
                       estimator=shrinkage_covariance_estimator,
-                      mass_weighted=True,
+                      weights=None,
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/encore/covariance.py]: This should be `"mass"` to retain the previous default behavior.

File: package/MDAnalysis/analysis/encore/covariance.py [hunk 6]
```diff
@@ -231,16 +224,27 @@ def covariance_matrix(ensemble,
...
         else:
-            masses = np.repeat(ensemble.atoms.masses, 3)
-
-        mass_matrix = np.sqrt(np.identity(len(masses))*masses)
-        sigma = np.dot(mass_matrix, np.dot(sigma, mass_matrix))
+            if selection:
+                req_len = ensemble.select_atoms(selection).n_atoms
+            else:
+                req_len = ensemble.atoms.n_atoms
+            if req_len != len(weights):
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/encore/covariance.py]: Isn't this going to fail when `weights == None`?

File: package/MDAnalysis/analysis/rms.py [hunk 7]
```diff
@@ -195,31 +195,25 @@ def rmsd(a, b, weights=None, center=False, superposition=False):
...
-        relative_weights = np.asarray(weights) / np.mean(weights)
-        relative_weights = relative_weights.astype(np.float64)
-    else:
-        relative_weights = None
-
     # superposition only works if structures are centered
     if center or superposition:
         # make copies (do not change the user data!)
         # weights=None is equivalent to all weights 1
         a = a - np.average(a, axis=0, weights=weights)
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/rms.py]: Not your PR but nicer would be
```python
a -= np.average(a, axis=0, weights=weights)
```

File: package/MDAnalysis/analysis/rms.py [hunk 8]
```diff
@@ -335,8 +330,10 @@ def __init__(self, atomgroup, reference=None, select='all',
 
         filename : str (optional)
             write RSMD into file file :meth:`RMSD.save`
-        mass_weighted : bool (optional)
+        mass_weighted : bool deprecated
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/rms.py]: (deprecated) ?

File: package/MDAnalysis/analysis/rms.py [hunk 9]
```diff
@@ -585,40 +567,27 @@ def run(self, start=None, stop=None, step=None, progout=10, quiet=False):
...
 
-        if not (rmsf >= 0).all():
+        if not (self.rmsf >= 0).all():
             raise ValueError("Some RMSF values negative; overflow " +
                              "or underflow occurred")
-
-        self._rmsf = rmsf
-
-    @property
-    def rmsf(self):
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/rms.py]: Having removed `RMSF.rmsf` should be mentioned in a versionchanged 0.16.0 doc, together with the transition to standard analysis API.

File: package/MDAnalysis/analysis/encore/covariance.py (hunk 6 above)
Comment (kain88-de) [inline package/MDAnalysis/analysis/encore/covariance.py]: This code path can only be reached if `if weights` above evaluates to true

File: package/MDAnalysis/analysis/rms.py (hunk 9 above)
Comment (kain88-de) [inline package/MDAnalysis/analysis/rms.py]: This isn't removed. It can be read the same way as before. Just that you can now also write to the value. So there isn't a change for the user. We handle all analysis classes right now that computed values can also be overwritten. So this is just making it behave more like the rest of MDAnalysis.

**Comment by kain88-de** [comment]
The final changes of this depend on #1145 to update encore as well to the new keyword.

File: package/MDAnalysis/analysis/encore/covariance.py (hunk 6 above)
Comment (orbeckst) [inline package/MDAnalysis/analysis/encore/covariance.py]: Ok

File: package/MDAnalysis/analysis/rms.py (hunk 9 above)
Comment (orbeckst) [inline package/MDAnalysis/analysis/rms.py]: Ok.

**Comment by orbeckst** [comment]
@kain88-de , I had some minor comments on the docs (please have a look) and the question of default values, to which you replied https://github.com/MDAnalysis/mdanalysis/pull/1136#issuecomment-271807514
> The final changes of this depend on #1145 to update encore as well to the new keyword.

Now that #1145 is done, what should the defaults for mass weighting be?

**Comment by kain88-de** [comment]
@orbeckst thanks for your comments. The default that I changed was to follow the function docs, if I remember correct. The default value for encore is a good question. @wouterboomsma and @mtiberti what weights would you prefer for encore as default masses or none? I'd appreciate input from you anyway with regard to the new feature to choose arbitrary weights.

**Comment by wouterboomsma** [comment]
@kain88-de Very nice work on supporting arbitrary weights. It should be easy to support this in the encore front-end as well. I think @mtiberti will just look into this. Regarding the default values, the current default is `mass_weighted=True`, so I guess we should set it to `weights='mass'` now?

**Comment by kain88-de** [comment]
Superseeded by #1197
